### ISSUE TITLE
Tabs/session lost after forced shutdown

### ISSUE THREAD

**Issue by gerhard4** [notepad-plus-plus/notepad-plus-plus#14781]
### Description of the Issue

This is now the second time that I have lost the session (that is, the open tabs) after a forced shutdown. 

### Steps to Reproduce the Issue

1. Open Notepad++ and open a few files.
2. Force a shutdown of the computer (long press on the power button on most laptops).
3. Boot the computer and open Notepad++.

### Expected Behavior

The prior session opens; that is, the same tabs are open as before.

### Actual Behavior

The session is empty; no files are open.

### Debug Information

```
Notepad++ v8.6.2   (64-bit)
Build time : Jan 14 2024 - 02:16:00
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 11 Pro (64-bit)
OS Version : 22H2
OS Build : 22621.3155
Current ANSI codepage : 1252
Plugins : 
    mimeTools (3)
    NppConverter (4.5)
    NppExport (0.4)
    NppPluginDemo (4.3)
```

**Comment by HenrikFFM** [comment]
same issue here:

`Notepad++ v8.6.2   (32-bit)
Build time : Jan 14 2024 - 02:18:41
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit)
OS Version : 22H2
OS Build : 19045.4046
Current ANSI codepage : 1252
Plugins : 
    ComparePlugin (2)
    DSpellCheck (1.4.6)
    mimeTools (3)
    NppConverter (4.5)
    NppExport (0.4)
`

**Comment by gerhard4** [comment]
I haven't looked at the code, but one suspicion is that the session file is not being closed after writing to it. Or at least flushed.

**Comment by xomx** [comment]
> 2\. Force a shutdown of the computer (long press on the power button on most laptops).

>  the session file is not being closed after writing to it. Or at least flushed.

Seriously? 

If this is not a trolling attempt - such a long press will cause the computer's ATX power supply to immediately cut off the power, so there will be no controlled shutdown of the Windows OS at all! So apparently no file writing or buffer flushing either. Just a power outage. Use this only as an emergency exit (e.g. for a totally frozen computer state...)

Notepad++ saves its session only at its process exit. From the v8.4.7+ it even handles well the Windows Update forced shutdowns/restarts. However, the Windows OS has to tell the running apps that such a shutdown is occurring and that they should save whatever is needed and quit. If you cut the power current - then everything will end as it is, immediately.

**Comment by gerhard4** [comment]
This is not a trolling attempt. This just happened. If you never had to kill a system, you're lucky. It happens.

> Notepad++ saves its session only at its process exit.

I don't know whether you assume this or know this from looking at the code. If Notepad++ wouldn't touch the session file before process exit, there wouldn't be a reason for it to be gone/empty after a forced shutdown -- it would contain the contents from the last shutdown. Which may not be the best, but in my situations it would have been what I needed. But that's not what happened -- the contents of the file were gone.

> So apparently no file writing or buffer flushing either.

It is quite possible to regularly update and flush or close a file such a session data, while the session is still ongoing, _before_ the process exits. You may lose the very latest update, but you're close and didn't lose _everything_. Many apps do this. Between my IDEs and editors, Notepad++ is unique in this scenario -- all others behave as expected.

**Comment by xomx** [comment]
@gerhard4 
> > Notepad++ saves its session only at its process exit.
> 
> I don't know whether you assume this or know this from looking at the code.

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/8bfe2d7a7748bf62f07d6a750215323f1be13c62/PowerEditor/src/NppBigSwitch.cpp#L2640-L2656

@donho 

Once you have already assigned yourself to this - we could probably update the session.xml file also at the end of each successful file-opening and closing in N++. What do you think?

**Comment by donho** [comment]
@xomx 
I assign to myself usually some critical issues as reminder.  But I would like to reproduce it later before treating it.

> You may lose the very latest update, but you're close and didn't lose everything. 

@gerhard4 is right. The session file should be untouched before writing it. IIRC, it's the case of Notepad++. So such situation cannot happen normally. Am I missing somethings?

**Comment by xomx** [comment]
@donho 

> After checking the code, `session.xml` is untouched by Notepad++ during its life - the only moment that Notepad++ rewrites it is on WM_CLOSE.

Agree.

> I'm wondering if Windows detects the incident and send WM_CLOSE to applications, and Notepad++ was trying to save the session but without enough time - that causes the empty file.

I do not think so. In that case, Windows cannot even write to its system log itself that such a power outage occurred, so why should it try to tell the running apps to attempt something themselves?

I rather think that OP launched N++ with no previous files opened, so the session.xml was simply empty when that power outage happened.

> If it's the case, we can do nothing about it.

No, we can. Like I write above (after considering the @gerhard4 's proposal) -  we could try to update the session.xml file also at the end of each successful file-opening and closing in N++.

**Comment by donho** [comment]
@xomx 
> I rather think that OP launched N++ with no previous files opened, so the session.xml was simply empty when that power outage happened.

Do you mean that 2 cases in this issue are under the situation that:
Notepad++ is launched with no file in the session (the 1st launched?), and after the power outage, restart the OS & Notepad++, all opened files in the last session are no more in the current session?

@gerhard4 & @HenrikFFM Could you confirm that above situation is your case?

**Comment by xomx** [comment]
@donho 
Exactly.

And my possible future enhancement proposal is about adding similar session.xml saving steps as is now in the WM_CLOSE, also before this line (for the successful file-opening in N++ case): https://github.com/notepad-plus-plus/notepad-plus-plus/blob/8bfe2d7a7748bf62f07d6a750215323f1be13c62/PowerEditor/src/ScintillaComponent/Buffer.cpp#L803

Then adding the same session.xml updating also for the successful file-closing case. That way if a power outage happens, the session.xml file list should be up to date and will not be so dependent on the WM_CLOSE handler, as it is now. IMO - more resiliency.

**Comment by donho** [comment]
@xomx 
> And my possible future enhancement proposal is about adding similar session.xml saving steps as is now in the WM_CLOSE, also before this line (for the successful file-opening in N++ case):

I was considering such solution. However, imagine if user loads 500 files, there will be 500 save operations - it's just not possible.

**Comment by gerhard4** [comment]
Thanks, guys, for looking into this.

> Notepad++ is launched with no file in the session (the 1st launched?), and after the power outage, restart the OS & Notepad++, all opened files in the last session are no more in the current session?

I had a long-lived session with about 10 files that had already survived several restarts of Notepad++ (and of the system). I'm reasonably certain that the session file had content before it got emptied.

> I'm wondering if Windows detects the incident and send WM_CLOSE to applications, and Notepad++ was trying to save the session but without enough time - that causes the empty file. If it's the case, we can do nothing about it.

I wondered about this myself, and this seems a reasonable scenario. It's also one that is dependent on certain timings and probably difficult to reproduce.

What I've seen done (and done myself on occasion) is to 
1. write the new file into a different filename (and close it), 
2. delete the current file, 
3. rename the just written file to the normal filename. 
 
When starting, the app 
1. checks for the normal file and uses it if it finds it,
2. when it doesn't find it, it checks for the "write" file with the different filename and uses it if it finds it
3. and only if it finds neither, it uses an empty session.

This guarantees that if a shutdown interrupts the process at any point in the write/delete/rename process, even during an operation, there is always either the old session file or the new session file, fully written and intact, present.

> However, imagine if user loads 500 files, there will be 500 save operations - it's just not possible.

How about checking the session for changes every 5 to 10 minutes (possibly configurable) and write if there are changes?

**Comment by xomx** [comment]
@gerhard4 

> * write the new file into a different filename (and close it),
> 
> * delete the current file,
> 
> * rename the just written file to the normal filename.

https://github.com/notepad-plus-plus/notepad-plus-plus/commit/50b81eadefa0aaeeece6e4458f45e9b931191191

> How about checking the session for changes every 5 to 10 minutes (possibly configurable) and write if there are changes?

I am not for it.

**Comment by gerhard4** [comment]
> https://github.com/notepad-plus-plus/notepad-plus-plus/commit/50b81eadefa0aaeeece6e4458f45e9b931191191: "... if written session file is corrupted ..."

This seems to address a different kind of problem. If an interruption of the write problem deletes or empties the session file, this may not help.

**Comment by donho** [comment]
@xomx 
> Hmm, it is not an efficient way as it is not a centralized way - there are more such file-opening places, e.g. by dragging - then there would have to be another place to add:

Yes. That's why it's not been done in this way. And I won't go for it. 


>>How about checking the session for changes every 5 to 10 minutes (possibly configurable) and write if there are changes?

>I am not for it.

I'm not for it either.

>> I'm wondering if Windows detects the incident and send WM_CLOSE to applications, and Notepad++ was trying to save the session but without enough time - that causes the empty file. If it's the case, we can do nothing about it.

> I wondered about this myself, and this seems a reasonable scenario. It's also one that is dependent on certain timings and probably difficult to reproduce.

I will test it this evening with my different laptop computer.
I'll let you guys know the test result.

**Comment by xomx** [comment]
@gerhard4 

> If an interruption of the write problem deletes or empties the session file, this may not help.

All possibilities one can end up with:
1) session.xml only (with correct previous content)
2) both session.xml (incorrect updated content) and session.xml.inCaseOfCorruption.bak (correct previous content)
3) both session.xml (correct updated content) and session.xml.inCaseOfCorruption.bak (correct previous content)
4) session.xml.inCaseOfCorruption.bak only (with correct previous content)
5) session.xml only (correct updated content)

@donho 

What is probably missing - when calling the `NppParameters::loadSession` and the session.xml file fails to load (that file is either missing or has invalid content), then we need to also look for its possible session.xml.inCaseOfCorruption.bak backup and use it instead.

**Comment by xomx** [comment]
@gerhard4 

> I don't know what "fails to load" captures, but a completely empty session.xml file could also be a scenario that triggers looking for the backup file.

It is covered, the completely missing file here:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/8bfe2d7a7748bf62f07d6a750215323f1be13c62/PowerEditor/src/Parameters.cpp#L2303
and its invalid (thus empty too) content here:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/8bfe2d7a7748bf62f07d6a750215323f1be13c62/PowerEditor/src/Parameters.cpp#L2305

**Comment by donho** [comment]
@gerhard4 @HenrikFFM 
BTW could you upload the file "%APPDATA%\Notepad++\nppLogNulContentCorruptionIssue.log" here please?

**Comment by gerhard4** [comment]
[nppLogNulContentCorruptionIssue.log](https://github.com/notepad-plus-plus/notepad-plus-plus/files/14464612/nppLogNulContentCorruptionIssue.log)

The recent forced shutdowns that prompted me to create this issue were this year and are not in this file.

**Comment by donho** [comment]
@gerhard4 
Thank you for the log.

> I will test it this evening with my different laptop computer.
I'll let you guys know the test result.


Tested with 3 laptops several times by holding power button to force Windows shutdown, with the Notepad++ (of PR #14823) opened. The cautious conditions for logging are removed in PR #14823, so any file writing is logged. 

I cannot reproduce the issue at all with the way to force shutdown Windows by holding power button - not only the session has always restored after restarting the PC and Notepad++, but also the writing session file doesn't happen (can't see any track in the log file) by forcing shutdown PC.

>  The session file should be untouched before writing it. IIRC, it's the case of Notepad++. 

I was wrong. During the life time of Notepad++, the session is saved every time when the opened file is modified (clean->dirty) or saved after being modified (dirty->clean), for saving the backup file state in session.xml.

Based on the above facts I described, I can imagine one scenario to reproduce such issue: 
When user modify a file at the first time (from clean to dirty) in Notepad++, and the time of periodic backup (7 sec by default) is reached: **the backup of modified file** and **session.xml** are being written sequentially - then the power outrage happens at this moment.

When above scenario happens, the corruption of file could be on **the backup of modified file** or **session.xml**, depend on the precise moment of  the power outrage. I believe it's the best explication also for [the NUL file-corruption bug](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/6133)

**Comment by donho** [comment]
@gerhard4 

Could you please check if you have such file: ***"%APPDATA%\Notepad++\session.xml.inCaseOfCorruption.bak"***?
If yes, could you please provide it here?

**Comment by gerhard4** [comment]
> Could you please check if you have such file: "%APPDATA%\Notepad++\session.xml.inCaseOfCorruption.bak"?

No such file.

**Comment by donho** [comment]
> No such file.

OK, thank you for checking.

**Comment by xomx** [comment]
@gerhard4 @HenrikFFM 

Could you please confirm that you do not have the Dell SupportAssist SW installed on your computers?

(check if you have the `C:\Program Files\Dell\\SupportAssistAgent\SRE\SRE.exe` on your disk, there is a [proof](https://community.notepad-plus-plus.org/post/94243) that this external program, in its latest version updated in January 2024, is currently deleting the Notepad++ `session.xml` file, probably because of it uses its own file with the same name...)

**Comment by gerhard4** [comment]
@xomx 
> Could you please confirm that you do not have the Dell SupportAssist SW installed on your computers?

Thank you! Yes, indeed I have this software installed. The post you linked is instructive, and sounds exactly like what I experienced -- and provides an explanation why this has never before been a problem.

**Comment by Kscheer71** [comment]
I see similar behaviour. I was used to Notepad++ automatically “saving” my changes in not saved tabs, so when I closed and open again the application what I wrote was there (really useful to use the application for long term notes). Since I installed Windows 11, this is not happening anymore. I checked the settings and it’s checked to remember current session for next launch. Notepadd++ is a virtualized one. Any idea what is the problem ?
